source/id_descriptor.h

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_ID_DESCRIPTOR_H_
+#define LIBSPIRV_ID_DESCRIPTOR_H_
+
+#include <unordered_map>
+#include <vector>
+
+#include "spirv-tools/libspirv.hpp"
+
+namespace libspirv {
+
+// Computes and stores id descriptors.
+//
+// Descriptors are computed as hash of all words in the instruction where ids
+// were substituted with previosly computed descriptors.
+class IdDescriptorCollection {
+ public:
+  IdDescriptorCollection() {
+    words_.reserve(16);
+  }
+
+  // Issues new descriptor for the result id of the given instruction.
+  // This function needs to be sequentially called for every instruction in the
+  // module.
+  uint32_t IssueNewDescriptor(const spv_parsed_instruction_t& inst);
```

Comment:
But it might be a reused descriptor.  So the "New" part of the name seems wrong.
Suggest "MakeDescriptorForInst" ?

Also say that the return value is the descriptor is associated with the instruction.